File: elixir/2015/day08a/Program.exs
#!/usr/bin/env elixir

defmodule Program do
  def main(args) do
    case args do
      [filename] ->
        case process(filename) do
          {:ok, result} ->
            IO.puts("result = #{result}")
          {:error, reason} ->
            IO.puts(:stderr, "error while processing file `#{filename}': #{reason}")
            System.halt(1)
        end
      _ ->
        usage()
    end
  end

  defp usage do
    IO.puts(:stderr, "usage: Program.exs <input file>")
    System.halt(1)
  end

  defp process(filename) do
    case File.read(filename) do
      {:ok, content} ->
        total_area =
          content
          |> String.split()
          |> Enum.reduce(0, fn
            line, acc ->
              code_len = String.length(line)
              quoted = String.slice(line, 1..(code_len - 2))
              mem_len = scan_quoted(0, quoted)
              acc + (code_len - mem_len)
          end)
        {:ok, total_area}
      {:error, reason} ->
        {:error, reason}
    end
  end

  defp scan_quoted(i, quoted) do
    quoted_len = String.length(quoted)
    if quoted_len == 0 do
      i
    else
      case String.at(quoted, 0) do
        "\\" ->
          case String.at(quoted, 1) do
            "\\" -> scan_quoted(i + 1, String.slice(quoted, 2..quoted_len))
            "\"" -> scan_quoted(i + 1, String.slice(quoted, 2..quoted_len))
            "x" -> scan_quoted(i + 1, String.slice(quoted, 4..quoted_len))
            _ -> scan_quoted(i + 1, String.slice(quoted, 1..quoted_len))
          end
        _ -> scan_quoted(i + 1, String.slice(quoted, 1..quoted_len))
      end
    end
  end
end

Program.main(System.argv())

File: erlang/2015/day08a/Program.erl
#!/usr/bin/env escript

main(Args) ->
  case Args of
    [Filename] ->
      case process(Filename) of
        {ok, Result} ->
          io:format("result = ~p~n", [Result]);
        {error, Reason} ->
          io:format(standard_error, "error while processing file `~s': ~p~n", [Filename, Reason]),
          halt(1)
      end;
    _ ->
      usage()
  end.

usage() ->
  io:format(standard_error, "usage: Program.erl <input file>~n", []),
  halt(1).

scan_quoted(Quoted, I) ->
  case Quoted of
    [] -> I;
    [_] -> I + 1;
    [Ch1, Ch2 | Rest] ->
      case {Ch1, Ch2} of
        {92, 34} -> scan_quoted(Rest, I + 1);
        {92, 92} -> scan_quoted(Rest, I + 1);
        {92, 120} -> scan_quoted(tl(tl(Rest)), I + 1);
        {_, _} -> scan_quoted([Ch2 | Rest], I + 1)
      end
  end.

process(Filename) ->
  case file:read_file(Filename) of
    {ok, Chars} ->
      Lines = string:lexemes(Chars, "\n"),
      Result = lists:foldl(
        fun (Line, Acc) ->
          CodeLen = string:length(Line),
          Quoted = string:slice(Line, 1, string:length(Line) - 2),
          MemLen = scan_quoted(string:to_graphemes(Quoted), 0),
          Acc + CodeLen - MemLen
        end,
        0,
        Lines),
      {ok, Result};
    {error, Reason} ->
      {error, Reason}
  end.
